app/server/lib/OIDCConfig.ts

```diff
@@ -0,0 +1,145 @@
+ * A good informative overview of SAML is at https://developer.okta.com/blog/2019/10/21/illustrated-guide-to-oauth-and-oidc
+ * Note:
+ *    SP is "Service Provider", in our case, the Grist application.
+ *    IdP is the "Identity Provider", somewhere users log into, e.g. Okta or Google Apps.
+ *
+ * We also use optional attributes for the user's name, for which we accept any of:
+ *    given_name
+ *    family_name
+ *
+ * Expected environment variables:
+ *    env GRIST_OIDC_SP_HOST=https://<your-domain>
+ *        Host at which our /oidc endpoint will live; identifies our application.
+ *    env GRIST_OIDC_IDP_ISSUER
+ *        The issuer URL for the IdP.
+ *    env GRIST_OIDC_IDP_CLIENT_ID
+ *        The client ID for the application, as registered with the IdP.
+ *    env GRIST_OIDC_IDP_CLIENT_SECRET
+ *        The client secret for the application, as registered with the IdP.
+ *    env GRIST_OIDC_IDP_SCOPES
+ *        The scopes to request from the IdP, as a space-separated list. Defaults to "openid email profile".
+ *    env GRIST_OIDC_IDP_CODE_CHALLENGE_METHOD
+ *        The method to use for code challenge. Defaults to "S256".
+ *
+ * This version of OIDCConfig has been tested with Keycloak OIDC IdP following the instructions
+ * at:
+ *   https://www.keycloak.org/getting-started/getting-started-docker
+ * When running on localhost and http, the settings tested were with:
+ *   - GRIST_OIDC_SP_HOST=http://localhost:8484 (or whatever port you use for Grist)
+ *   - GRIST_OIDC_IDP_ISSUER=http://localhost:8080/realms/myrealm (replace 8080 by the port you use for keycloak)
+ *   - GRIST_OIDC_IDP_CLIENT_ID=myclient
+ *   - GRIST_OIDC_IDP_CLIENT_SECRET=YOUR_SECRET (as set in keycloak)
+ *   - GRIST_OIDC_IDP_SCOPES="openid email profile"
+ */
+
+import * as express from 'express';
+import {GristLoginSystem, GristServer} from './GristServer';
+import { Client, generators, Issuer } from 'openid-client';
+import {Sessions} from './Sessions';
+const CALLBACK_URL = '/oauth2/callback';
+const LOGIN_URL = '/oauth2/login';
+
+export class OIDCConfig {
+  private _client: Client;
+  private _nonce: string;
+  private _codeVerifier: string;
+  private _codeChallenge: string;
+
+  public constructor() {
+    this._nonce = generators.nonce();
+    this._codeVerifier = generators.codeVerifier();
+    this._codeChallenge = generators.codeChallenge(this._codeVerifier);
+  }
+
+  public async initOIDC(): Promise<void> {
+    if (!process.env.GRIST_OIDC_SP_HOST) { throw new Error('initConfig requires GRIST_OIDC_SP_HOST to be set'); }
+    if (!process.env.GRIST_OIDC_IDP_ISSUER) { throw new Error('initConfig requires GRIST_OIDC_IDP_ISSUER to be set'); }
+    if (!process.env.GRIST_OIDC_IDP_CLIENT_ID) {
+      throw new Error('initConfig requires GRIST_OIDC_IDP_CLIENT_ID to be set');
+    }
+    if (!process.env.GRIST_OIDC_IDP_CLIENT_SECRET) {
+      throw new Error('initConfig requires GRIST_OIDC_IDP_CLIENT_SECRET to be set');
+    }
```

Comment:
Nit: probably should be `initOIDC` instead of `initConfig` in these error messages?